src/compiler/crystal/semantic/type_declaration_processor.cr

```diff
@@ -758,9 +757,9 @@ struct Crystal::TypeDeclarationProcessor
     end
   end
 
-  private def uninstantiate(type)
+  private def uninstantiate(type) : Type
     if type.is_a?(GenericInstanceType)
-      type.generic_type
+      type.generic_type.as(Type)
```

Comment:
`GenericInstanceType#generic_type` returns `GenericType` hence the resulting type would be a union `GenericType | Type`. All including types of `GenericType` inherit from `Type` but the compiler doesn't merge this union.